src/Stratis.Bitcoin.Features.PoA/Voting/JoinFederationRequestService.cs

```diff
@@ -0,0 +1,102 @@
+using Stratis.Features.PoA.Collateral.CounterChain;
+
+namespace Stratis.Features.PoA.Voting
+{
+    public interface IJoinFederationRequestService
+    {
+        Task<PubKey> JoinFederationAsync(JoinFederationRequestModel request, CancellationToken cancellationToken);
+    }
+
+    public sealed class JoinFederationRequestService : IJoinFederationRequestService
+    {
+        private readonly ICounterChainSettings counterChainSettings;
+        private readonly IFullNode fullNode;
+        private readonly IHttpClientFactory httpClientFactory;
+        private readonly ILoggerFactory loggerFactory;
+        private readonly PoANetwork network;
+        private readonly NodeSettings nodeSettings;
+        private readonly VotingManager votingManager;
+
+        public JoinFederationRequestService(ICounterChainSettings counterChainSettings, IFullNode fullNode, IHttpClientFactory httpClientFactory, ILoggerFactory loggerFactory, Network network, NodeSettings nodeSettings, VotingManager votingManager)
+        {
+            this.counterChainSettings = counterChainSettings;
+            this.fullNode = fullNode;
+            this.httpClientFactory = httpClientFactory;
+            this.loggerFactory = loggerFactory;
+            this.network = network as PoANetwork;
+            this.nodeSettings = nodeSettings;
+            this.votingManager = votingManager;
+        }
+
+        public async Task<PubKey> JoinFederationAsync(JoinFederationRequestModel request, CancellationToken cancellationToken)
+        {
+            // Get the address pub key hash.
+            BitcoinAddress address = BitcoinAddress.Create(request.CollateralAddress, this.counterChainSettings.CounterChainNetwork);
+            KeyId addressKey = PayToPubkeyHashTemplate.Instance.ExtractScriptPubKeyParameters(address.ScriptPubKey);
+
+            // Get mining key.
+            var keyTool = new KeyTool(this.nodeSettings.DataFolder);
+            Key minerKey = keyTool.LoadPrivateKey();
+            if (minerKey == null)
+                throw new Exception($"The private key file ({KeyTool.KeyFileDefaultName}) has not been configured or is not present.");
+
+            var expectedCollateralAmount = CollateralFederationMember.GetCollateralAmountForPubKey(this.network, minerKey.PubKey);
+
+            var collateralAmount = new Money(expectedCollateralAmount, MoneyUnit.BTC);
+
+            var joinRequest = new JoinFederationRequest(minerKey.PubKey, collateralAmount, addressKey);
+
+            // Populate the RemovalEventId.
+            var collateralFederationMember = new CollateralFederationMember(minerKey.PubKey, false, joinRequest.CollateralAmount, request.CollateralAddress);
+
+            byte[] federationMemberBytes = (this.network.Consensus.ConsensusFactory as CollateralPoAConsensusFactory).SerializeFederationMember(collateralFederationMember);
+            Poll poll = this.votingManager.GetFinishedPolls().FirstOrDefault(x => x.IsExecuted &&
```

Comment:
Perhaps change this to order by poll height and take the last match with `LastOrDefault`.